core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java

```diff
@@ -0,0 +1,445 @@
+package io.substrait.relation;
+
+import io.substrait.expression.Expression;
+import io.substrait.expression.ExpressionVisitor;
+import io.substrait.expression.FieldReference;
+import io.substrait.expression.ImmutableFieldReference;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+/**
+ * Class used to visit all child relations from a root relation and optionally replace subtrees by
+ * overriding a visitor method. The traversal will include relations inside of subquery expressions.
+ * By default, no subtree substitution will be performed. However, if a visit method is overridden
+ * to return a different relation than the one passed in, then that relation will replace it in the
+ * tree.
+ */
+public class RelCopyOnWriteVisitor implements RelVisitor<Rel, RuntimeException> {
+  @Override
+  public Aggregate visit(Aggregate aggregate) {
```

Comment:
When there's a single input like above, it looks pretty clean, but when there are multiple potentially changing arguments to the builder, it gets uglier with the `orElse` calls all over IMHO.

Reply:
Yeah, the multiples are a mix. I agree. A utility can clean up the part of this (the latter part of the function). Maybe something like this:

```
// this uses a variable
public T Optional<T> compose(Supplier<T> supplier, Optional... optionals) {
  for (Optional<?> optional : optionals) {
    if (optional.isPresent()) {
      return supplier.get();
    }
  }

  return Optional.empty();
}
```

This cleans things up some. 

I actually seriously considered introducing a class something like `OptionalWithDefault<T>` that holds the original value or the alternative. Could be similarly composable but avoids the excess elsewhere.

Reply:
Having `OptionalWithDefault<T>` is interesting. It'd contain an `Optional<T>` and a `T`. Would have to define methods on it like `map` that delegate to the `Optional<T>`. But then getting the underlying value is nice and clean.

Reply:
Do you think that's worth doing? Then all methods would return `OptionalWithDefault<T>`. Might be useful down the road if we want a single object that has the old and new value for doing things like diffing.

Reply:
I'd suggest doing it on one or two methods and see how it pans out. It feels like it might make things much cleaner (when combined with my proposed compose function above).

Reply:
Here's an example with a few methods implemented. WDYT, @jacques-n ?
```
  public <T> OptionalWithDefault<T> compose(
      Supplier<T> supplier, T fallbackValue, OptionalWithDefault<?>... optionals) {
    for (OptionalWithDefault<?> optional : optionals) {
      if (optional.isPresent()) {
        return OptionalWithDefault.of(supplier.get(), fallbackValue);
      }
    }

    return OptionalWithDefault.of(fallbackValue);
  }

  @Override
  public OptionalWithDefault<Rel> visit(Aggregate aggregate) throws RuntimeException {
    return aggregate
        .getInput()
        .accept(this)
        .map(input -> ImmutableAggregate.builder().from(aggregate).input(input).build());
  }

  @Override
  public OptionalWithDefault<Rel> visit(Filter filter) throws RuntimeException {
    var input = filter.getInput().accept(this);
    var condition = visitExpression(filter.getCondition());
    return compose(
        () ->
            ImmutableFilter.builder()
                .from(filter)
                .input(input.get())
                .condition(condition.get())
                .build(),
        filter,
        input,
        condition);
  }
```

Reply:
Filter looks pretty good. What do you think?

Reply:
Looks pretty good. Here's on variant that chains a `withDefault` method on the return of `compose` so you don't need to pass through `filter` as an arg. Do you think that's better?
```
  @Override
  public OptionalWithDefault<Rel> visit(Filter filter) throws RuntimeException {
    var input = filter.getInput().accept(this);
    var condition = visitExpression(filter.getCondition());
    return compose(
            () ->
                ImmutableFilter.builder()
                    .from(filter)
                    .input(input.get())
                    .condition(condition.get())
                    .build(),
            input,
            condition)
        .withDefault(filter);
  }
```

Reply:
Yeah, nice! Maybe even can do instead:
```
OptionalWithDefault.ofDefault(filter)
  .onNonDefault(input, condition)
  .replaceWith(your supplier);
```
Or something like that. Better orders the arguments I think (and avoids the positional problem of the current impl).

Reply:
Settled on this pattern for now
```
  @Override
  public OptionalWithDefault<Rel> visit(Fetch fetch) throws RuntimeException {
    return fetch
        .getInput()
        .accept(this)
        .map(input -> ImmutableFetch.builder().from(fetch).input(input).build())
        .withDefault(fetch);
  }

  @Override
  public OptionalWithDefault<Rel> visit(Filter filter) throws RuntimeException {
    var input = filter.getInput().accept(this);
    var condition = visitExpression(filter.getCondition());
    return OptionalWithDefault.of(filter)
        .ifAnyPresent(input, condition)
        .overrideWith(
            () ->
                ImmutableFilter.builder()
                    .from(filter)
                    .input(input.get())
                    .condition(condition.get())
                    .build());
  }
```

Reply:
I'm finding that it's working better if the fallback returns an OptionalWithDefault with the default value set to the fallback argument. Otherwise, I have to add it myself everywhere manually.

Reply:
Getting ugly using `OptionalWithDefault`. It becomes too high a burden on the function, and if you don't make assumptions on what the caller returns, then setting up the default value becomes ugly. I think I'm going to switch to `Optional` with a couple of utility functions.